### ISSUE TITLE
'nodetool removenode' doesn't guarantee that the removed node doesn't get back to cluster

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#11217]
HEAD: a03a33dcaf08c99ce1bcf6023c1a52b855288ac0

*Description*
It's known that if you want to remove a node from a cluster you should use `decommission` if a removed node is alive and `removenode` if it's not.
But what do you need to do if you lost a connection to one or more nodes of your cluster and you have no clue if you are ever going to get it back?

The procedure https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/remove-node.html has the following:

-------------------------
* Using nodetool removenode is a fallback procedure that should only be used when a node is permanently down and cannot be recovered.
* You must never use nodetool removenode to remove a running node that can be reached by other nodes in the cluster.
--------------------------

After reading the above I'd continue to `removenode`. However a more cautious user would have read the https://docs.scylladb.com/stable/operating-scylla/nodetool-commands/removenode first.

There we have the following warning:

--------------------------
* You must never use the nodetool removenode command to remove a running node that can be reached by other nodes in the cluster. Before using the command, make sure the node is permanently down and cannot be recovered.

* If the node is up and reachable by other nodes, use nodetool decommission. See [Remove a Node from a Scylla Cluster](https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/remove-node.html) for more information.
--------------------------

The above again suggests that if node is up and reachable - use `decommission`. The only other option is using a `removenode`.

The bottom line: after reading our documentation a user would run `nodetool removenode` in case of the outage I described at the top.

But what if after that the outage is resolved and supposedly unreachable nodes become reachable again?
A naive user would assume that after `removenode` those nodes would not be allowed into the cluster somehow but he/she couldn't be more wrong. - Those nodes would simply "appear" as if there wasn't any `removenode` operation at all (no streaming either).

To reproduce:

1) Create a cluster of 3 nodes.
2) Stop two Scylla nodes.
3) Run `nodetool removenode <UUID of the node>` for each node from (2).
4) Start Scylla nodes from (2).

As a result the nodes from (2) is going to serve old data and all data written while they were down to new replicas is going to be lost.

**Comment by vladzcloudius** [comment]
One way to avoid this is to block the node-to-be-removed on a networking level on each alive Scylla node *before* you run `nodetool removenode`.

**Comment by vladzcloudius** [comment]
@tzach FYI. Please, consider at least updating our documentation about this "phenomena".

**Comment by xemul** [comment]
@vladzcloudius , sounds like the node should be [fenced](https://en.wikipedia.org/wiki/Fencing_(computing))

**Comment by vladzcloudius** [comment]
> @vladzcloudius , sounds like the node should be [fenced](https://en.wikipedia.org/wiki/Fencing_(computing))

Yup. Something like that.

**Comment by fee-mendes** [comment]
Reading through it, I believe it is quite similar as https://github.com/scylladb/scylladb/issues/10792. Different operation, but same scenario, a node thought to be gone gets back, mysterious things start to happen.

**Comment by vladzcloudius** [comment]
> Reading through it, I believe it is quite similar as #10792. Different operation, but same scenario, a node thought to be gone gets back, mysterious things start to happen.

Possibly. And I suspect that the "fix" for #10792 was adding the corresponding "harsh warning" at https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/replace-dead-node.html:

```
It’s essential to ensure the replaced (dead) node will never come back to the cluster, which might lead to a split-brain situation. Remove the replaced (dead) node from the cluster network or VPC.
```

**Comment by tzach** [comment]
> One way to avoid this is to block the node-to-be-removed on a networking level on each alive Scylla node before you run nodetool removenode.

@vladzcloudius this look like hard to do and error prune process.
Agree with @xemul that fencing or blacklisting the node in Scylla make sense.

AFAIK, this problem will be solved as part of moving to Raft base topology updates, which is very close.
@kostja is it the case?

**Comment by vladzcloudius** [comment]
> > One way to avoid this is to block the node-to-be-removed on a networking level on each alive Scylla node before you run nodetool removenode.
> 
> @vladzcloudius this look like hard to do and error prune process. 

I think so too - but this is what our Eng (e.g. @slivne) suggested to do when I filed this issue.

> Agree with @xemul that fencing or blacklisting the node in Scylla make sense.
> 
> AFAIK, this problem will be solved as part of moving to Raft base topology updates, which is very close. @kostja is it the case?

This issue has nothing to do with Raft nor Raft is going to help to resolve it on its own, @tzach.

You, guys, need to stop thinking that Raft is going to "resolve everything" (I don't remember who started this nonsense).

Raft is just a protocol that allows applying changes to a state of a distributed system in an orderly manner. It's like a semaphore if you'd like.

But we were missing here was a protocol (some handshake maybe) that would not allow a removed node back into the cluster. And this is a higher level problem than Raft - which is just a low level sync protocol.

I see that while you were waiting for Kostja to respond Asias sent and Avi merged a fix.

Thanks, guys. I couldn't ask for more!

**Comment by kbr-** [comment]
> You, guys, need to stop thinking that Raft is going to "resolve everything" (I don't remember who started this nonsense).

> Raft is just a protocol that allows applying changes to a state of a distributed system in an orderly manner. It's like a semaphore if you'd like.

> But what we were missing here was a protocol (some handshake maybe) that would not allow a removed node back into the cluster. And this is a higher level problem than Raft - which is just a low level sync protocol.

You're right that Raft on its own doesn't solve anything. It's how you use it determines what problems it will solve.

In our case, the plan is to move the topology state - including the token ring - into a Raft-replicated state machine that covers all nodes. When we remove a node from the cluster, using `removenode` for example, we will also remove it from this Raft group. Thus it won't be able to readd its tokens - other members of the group won't accept its attempt to do so.

**Comment by vladzcloudius** [comment]
> > You, guys, need to stop thinking that Raft is going to "resolve everything" (I don't remember who started this nonsense).
> 
> > Raft is just a protocol that allows applying changes to a state of a distributed system in an orderly manner. It's like a semaphore if you'd like.
> 
> > But what we were missing here was a protocol (some handshake maybe) that would not allow a removed node back into the cluster. And this is a higher level problem than Raft - which is just a low level sync protocol.
> 
> You're right that Raft on its own doesn't solve anything. It's how you use it determines what problems it will solve.
> 
> In our case, the plan is to move the topology state - including the token ring - into a Raft-replicated state machine that covers all nodes. When we remove a node from the cluster, using `removenode` for example, we will also remove it from this Raft group. Thus it won't be able to readd its tokens - other members of the group won't accept its attempt to do so.

Well, good thing is that there are other ways to prevent a node from bulging back into the cluster except for Raft groups and we don't need to wait for it to become GA in order to fix this bug. ;) 

Just like with Raft groups it starts and ends with a node to ASK to be accepted into a cluster - Raft or not Raft. And this part was missing. After https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8 - it is not missing anymore.

**Comment by kbr-** [comment]
> After https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8 - it is not missing anymore.

Well, that commit covers for restarting nodes that have been removed, but what if you removed a node that was merely partitioned and never shutdown? Once the network partition resolves, it will happily gossip with the cluster and reintroduce its tokens.

Not saying the patch doesn't improve the situation - it does for nodes that have been shutdown and removed - but be aware that it doesn't solve this issue completely.

We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated.

**Comment by kostja** [comment]
* Kamil Braun ***@***.***> [22/08/12 23:24]:
> We could of course implement a fencing/blacklisting mechanism
> independently of Raft to solve the problem (but the question is
> if it's worth it). If not, then I guess documentation still
> needs to be updated.

Such mechanism to do properly would require a linearizable
persistent storage to preserve its state.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by vladzcloudius** [comment]
> > After [12ab2c3](https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8) - it is not missing anymore.
> 
> Well, that commit covers for restarting nodes that have been removed, but what if you removed a node that was merely partitioned and never shutdown? Once the network partition resolves, it will happily gossip with the cluster and reintroduce its tokens.
> 
> Not saying the patch doesn't improve the situation - it does for nodes that have been shutdown and removed - but be aware that it doesn't solve this issue completely.
> 
> We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated.

Hmmm... I didn't realize that this patch only covers a bootup phase and doesn't make gossip ALWAYS check if "this" node is part of the ring, @asias.

Yes, it's worth fixing it NOW because this is a terrible bug, @kbr- .

Documenting that you have to configure firewalls on 120 nodes in order to prevent 30 other nodes from the partitioned AZ (that have been removed) from joining the cluster back is going to sound less than ridiculous.

**Comment by vladzcloudius** [comment]
> * Kamil Braun ***@***.***> [22/08/12 23:24]:
> We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated.
> Such mechanism to do properly would require a linearizable persistent storage to preserve its state.
> […](#)
> -- Konstantin Osipov, Moscow, Russia

@kostja in the case in question the state change has DEFINITELY took place on all alive nodes. The issue in question happens AFTER `removenode` DEFINITELY completed long ago and the corresponding side effect is visible to all alive nodes.

**Comment by vladzcloudius** [comment]
You, guys, seem to forget that removing the node or bootstrapping a node (in case of a "replace node") is taking an "infinite" amount of time. And the event that you care to "notice" is a beginning of such a process. 

So, you don't really need Raft to prevent a removed node from re-appearing - you simply need a gossiper to always make sure that the ring state of the local node and the rest of the cluster is the same and if it's not - make the local node shut itself down.

This protocol doesn't exist today AFAIU. And Raft could only make it more robust - but not replace it.

**Comment by vladzcloudius** [comment]
And from the single node prespective, @kostja , our storage is both linearizable and persistent ;)

**Comment by kostja** [comment]
* Vladislav Zolotarov ***@***.***> [22/08/13 00:11]:
> > We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated.
> > Such mechanism to do properly would require a linearizable persistent storage to preserve its state.
> > […](#)
> > -- Konstantin Osipov, Moscow, Russia
> 
> @kostja in the case in question the state change has DEFINITELY
> took place on all alive nodes. The issue in question happens
> AFTER `removenode` DEFINITELY completed long ago and the
> corresponding side effect is visible to all alive nodes.

I understand that from the external observer stand point it may be
true. But a distributed database doesn't have the luxury of having
a single view point and even different nodes are not guaranteed to
preserve its own opinion over time.

With raft the issue is easy to fix. I asked @gusev-p to look at
it.

Meanwhile we should stop pushing some partial fixes for non-raft
branch turning our life with supporting two fixes for the same
problem into a hell.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by avikivity** [comment]
Was this fixed by 12ab2c3d8d6925a14749be61b90a3032c994f1f8? If not, what're remaining?

This looks like it needs backporting, in general patches that fix importand bugs should Fix #foo instead of just #ref, or they miss getting backported.

**Comment by asias** [comment]
> * Vladislav Zolotarov ***@***.***> [22/08/13 00:11]:
> > We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated. > Such mechanism to do properly would require a linearizable persistent storage to preserve its state. > […](#) > -- Konstantin Osipov, Moscow, Russia @kostja in the case in question the state change has DEFINITELY took place on all alive nodes. The issue in question happens AFTER `removenode` DEFINITELY completed long ago and the corresponding side effect is visible to all alive nodes.
> I understand that from the external observer stand point it may be true. But a distributed database doesn't have the luxury of having a single view point and even different nodes are not guaranteed to preserve its own opinion over time. With raft the issue is easy to fix. I asked @gusev-p to look at it. Meanwhile we should stop pushing some partial fixes for non-raft branch turning our life with supporting two fixes for the same problem into a hell.

We do not need raft to protect removed nodes to rejoin the cluster. 

By "stop pushing some partial fixes non-raft branch", are you suggesting we should revert the fix https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8 and wait for raft to fix everything?

BTW, I have a plan to fix the harder cases where network partition recover caused rejoin too. 

https://github.com/scylladb/scylladb/pull/11244#issuecomment-1214521169

So, again, without raft, we can have a robust solution too.

**Comment by asias** [comment]
> Was this fixed by [12ab2c3](https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8)? If not, what're remaining?

Avi, See here: 

https://github.com/scylladb/scylladb/pull/11244#issuecomment-1214521169
> 
> This looks like it needs backporting, in general patches that fix importand bugs should Fix #foo instead of just #ref, or they miss getting backported.

The fix 12ab2c3d8d6925a14749be61b90a3032c994f1f8 uses refs instead of fixes because it only fixes part of the problem.

**Comment by kostja** [comment]
* Asias He ***@***.***> [22/08/15 10:02]:
> > * Vladislav Zolotarov ***@***.***> [22/08/13 00:11]:
> > > We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated. > Such mechanism to do properly would require a linearizable persistent storage to preserve its state. > […](#) > -- Konstantin Osipov, Moscow, Russia @kostja in the case in question the state change has DEFINITELY took place on all alive nodes. The issue in question happens AFTER `removenode` DEFINITELY completed long ago and the corresponding side effect is visible to all alive nodes.
> > I understand that from the external observer stand point it may be true. But a distributed database doesn't have the luxury of having a single view point and even different nodes are not guaranteed to preserve its own opinion over time. With raft the issue is easy to fix. I asked @gusev-p to look at it. Meanwhile we should stop pushing some partial fixes for non-raft branch turning our life with supporting two fixes for the same problem into a hell.
> 
> We do not need raft to protect removed nodes to rejoin the cluster. 
> 
> By "stop pushing some partial fixes non-raft branch", are you suggesting we should revert the fix https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8 and wait for raft to fix everything?
> 
> BTW, I have a plan to fix the harder cases where network partition recover caused rejoin too. 
> 
> https://github.com/scylladb/scylladb/pull/11244#issuecomment-1214521169
> 
> So, again, without raft, we can have a robust solution too. 

Yes, we do. But please don't, since we already settled on raft.


-- 
Konstantin Osipov, Moscow, Russia

**Comment by vladzcloudius** [comment]
> * Asias He ***@***.***> [22/08/15 10:02]:
> > * Vladislav Zolotarov ***@***.***> [22/08/13 00:11]: > > We could of course implement a fencing/blacklisting mechanism independently of Raft to solve the problem (but the question is if it's worth it). If not, then I guess documentation still needs to be updated. > Such mechanism to do properly would require a linearizable persistent storage to preserve its state. > […](#) > -- Konstantin Osipov, Moscow, Russia @kostja in the case in question the state change has DEFINITELY took place on all alive nodes. The issue in question happens AFTER `removenode` DEFINITELY completed long ago and the corresponding side effect is visible to all alive nodes. > I understand that from the external observer stand point it may be true. But a distributed database doesn't have the luxury of having a single view point and even different nodes are not guaranteed to preserve its own opinion over time. With raft the issue is easy to fix. I asked @gusev-p to look at it. Meanwhile we should stop pushing some partial fixes for non-raft branch turning our life with supporting two fixes for the same problem into a hell. We do not need raft to protect removed nodes to rejoin the cluster. By "stop pushing some partial fixes non-raft branch", are you suggesting we should revert the fix [12ab2c3](https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8) and wait for raft to fix everything? BTW, I have a plan to fix the harder cases where network partition recover caused rejoin too. [#11244 (comment)](https://github.com/scylladb/scylladb/pull/11244#issuecomment-1214521169) So, again, without raft, we can have a robust solution too.
> Yes, we do. But please don't, since we already settled on raft.
> […](#)
> -- Konstantin Osipov, Moscow, Russia

@kostja As much as I understand where you are coming from unless we are going to have a GA Raft-based solution that is backportable (!!!!)  to our Enterprise builds (based on 4.6 (2021.1.x) and 5.0 (2022.1) OSS versions) in a matter of single digit weeks from now I don't see how we can afford following your advice.

And AFAIK we are NOT going to have the above condition satisfied.

**Comment by mykaul** [comment]
@asias - what's the status of your fix (mentioned @ https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214530454 ) ?

**Comment by asias** [comment]
> @asias - what's the status of your fix (mentioned @ [#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214530454) ) ?

@mykaul  I was asked not to work on this: https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214873604 and https://github.com/scylladb/scylla-enterprise/issues/2338#issuecomment-1203609209.

I think it is wrong to wait for raft for everything, so I worked out a PoC anyway, in my spare time.

**Comment by asias** [comment]
> > @asias - what's the status of your fix (mentioned @ [#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214530454) ) ?
> 
> @mykaul I was asked not to work on this: [#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214873604) and [scylladb/scylla-enterprise#2338 (comment)](https://github.com/scylladb/scylla-enterprise/issues/2338#issuecomment-1203609209).
> 
> I think it is wrong to wait for raft for everything, so I worked out a PoC anyway, in my spare time.

Here is the PoC: 

https://github.com/scylladb/scylladb/pull/12040

**Comment by kostja** [comment]
* Asias He ***@***.***> [22/11/21 04:57]:
> > > @asias - what's the status of your fix (mentioned @ [#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214530454) ) ?
> > 
> > @mykaul I was asked not to work on this: [#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214873604) and [scylladb/scylla-enterprise#2338 (comment)](https://github.com/scylladb/scylla-enterprise/issues/2338#issuecomment-1203609209).
> > 
> > I think it is wrong to wait for raft for everything, so I worked out a PoC anyway, in my spare time.

I don't understand what it means "wait for raft". Raft is in. What
are you waiting for now?

> Here is the PoC: 
> 
> https://github.com/scylladb/scylladb/pull/12040

-- 
Konstantin Osipov, Moscow, Russia

**Comment by asias** [comment]
> * Asias He ***@***.***> [22/11/21 04:57]:
> > > @asias - what's the status of your fix (mentioned @ [#11217 (comment)]([#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214530454)) ) ? > > @mykaul I was asked not to work on this: [#11217 (comment)]([#11217 (comment)](https://github.com/scylladb/scylladb/issues/11217#issuecomment-1214873604)) and [[scylladb/scylla-enterprise#2338](https://github.com/scylladb/scylla-enterprise/issues/2338) (comment)]([scylladb/scylla-enterprise#2338 (comment)](https://github.com/scylladb/scylla-enterprise/issues/2338#issuecomment-1203609209)). > > I think it is wrong to wait for raft for everything, so I worked out a PoC anyway, in my spare time.
> I don't understand what it means "wait for raft". Raft is in. What are you waiting for now?

Wait for raft = Depend on raft. 

Also, please see Vlad's comment:

https://github.com/scylladb/scylladb/issues/11217#issuecomment-1217166656

**Comment by mykaul** [comment]
I'm unsure where we stand with this, but certainly unsure we should consider backporting it right now, as it's open.

**Comment by kostja** [comment]
It is part of the new topology changes and will be addressed by @gusev-p patch on fencing. The fencing spec has got consensus last week. I think we'll have it in experimental mode in 5.3 (although it's hard to commit since the patch is not submitted yet).

**Comment by kostja** [comment]
The backport label definitely needs to be removed, it's part of a huge feature we will not be able to backport.

**Comment by kbr-scylla** [comment]
Ref: https://github.com/scylladb/scylladb/issues/16108

**Comment by kbr-scylla** [comment]
@gleb-cloudius please confirm that your new patchset on the mailing list fixes this too. (It bans node at the beginning of removenode, IIUC)

**Comment by gleb-cloudius** [comment]
> @gleb-cloudius please confirm that your new patchset on the mailing list fixes this too. (It bans node at the beginning of removenode, IIUC)

This issue (as described in the opening message) was fixed with raft topology when we started to ban nodes in the "left" state. What my patchset will fix is the case where a node re-appears in the middle of the removenode operation.

**Comment by kbr-scylla** [comment]
This is fixed by 7e9e10186f1880f7e711b7363133e5e9947204ff in raft-topology mode, for 6.0.
